src/conn/conn_dhandle.c

```diff
@@ -134,7 +134,17 @@ __conn_dhandle_mark_dead(WT_SESSION_IMPL *session)
 	 * eviction and drain any blocks already queued.
 	 */
 	WT_RET(__wt_evict_file_exclusive_on(session, &evict_reset));
+
+	/* The handle is dead. */
 	F_SET(session->dhandle, WT_DHANDLE_DEAD);
+
+	/*
+	 * Reset the tree's eviction priority, and the tree is evictable by
+	 * definition.
+	 */
+	__wt_evict_priority_clear(session);
+	F_CLR(S2BT(session), WT_BTREE_NO_EVICTION);
```

Comment:
@agorrod
> the current behavior is expected - sometimes we want to clear WT_BTREE_NO_EVICTION so that eviction keeps looking at the file and sometimes we are planning to flush the file ourselves and need eviction to be locked out.

First, I'm sorry, but I still don't understand your point.

There are four cases where we use`__wt_evict_file_exclusive_on/__wt_evict_file_exclusive_off` pairs:

*  `__sync_file: WT_SYNC_CHECKPOINT`: doesn't want to lock out eviction, just doing an on/off pair to drain the queue as part of an eviction state change,
* `__conn_dhandle_mark_dead`: doesn't want to lock out eviction, doing an off/off pair to drain the queue as part of an eviction state change,
* `__sweep_expire_one`: does want to lock out eviction,
* `__wt_evict_file`: does want to lock out eviction

The code looks correct to me, those functions call `__wt_evict_file_exclusive_on`, then call `__wt_evict_file_exclusive_off` to clear `WT_BTREE_NO_EVICTION` once the exclusive operation is complete.

Second, I don't think this branch changes the previous behavior, so any problem you're seeing is probably in develop as well.